### ISSUE TITLE
Changing ProductType of Product - what problems can it cause? Attributes of ProductVariant

### ISSUE THREAD

**Issue by Boquete** [mirumee/saleor#5227]
Right now, we cannot easly change ProductType of Product in dashboard etc. so there must be a reason behind it.

Why it's blocked?
What if we need to migrate many products from one ProductType to another simmiliar one with different attributes, what problems can it make?

Also - why ProductVariant attributes can't be optional? Multiple product shares same ProductType, but not all variants uses all attributes. It was possible in past to leave empty values but not now. Any reason behind that?

**Comment by maarcingebala** [comment]
Changing the product type of an already existing is not supported yet. Product type is a template of a product - it defines some of its fields. If you changed the product type, we'd have to clear already assigned attributes, rerender the dashboard's product page to include new ones and "force" the user to fill them (as some may be required).

It's easy for product attributes but a bit harder for variant attributes (which brings us to your second question). Currently, storefront relies on variant attributes in the variant picker components:

![image](https://user-images.githubusercontent.com/5421321/73922632-c80ad180-48c9-11ea-9bf1-0d14bcca8e46.png)

To add a variant to the cart, you need to pick values of all its attributes so that we can identify a particular variant instance.

These are all just our assumptions that have existed in Saleor for a long time before we started extending our attributes system with additional parameters (such as attribute value type, visibility in storefront etc). I think now the entire attributes model would be more ready to support having optional variant attributes, but it's simply not implemented yet. The same goes for changing the product type - it would be also possible to add, but it would require a nice UI to let users easily "migrate" their products without breaking them.

I need to mention here one related problem that we have now - you can currently change the product type itself by adding new variant attributes, but this brokes products that already use that type. This is a similar case where we should allow the users to migrate existing products to their new definitions.

I hope this explains the situation a bit ;)

**Comment by Boquete** [comment]
> I need to mention here one related problem that we have now - you can currently change the product type itself by adding new variant attributes, but this brokes products that already use that type. This is a similar case where we should allow the users to migrate existing products to their new definitions.

@maarcingebala  right now admin user, can change attributes of product type and most important *assign new variant attributes* to existing product type. Problem with that is - that you would need to assign those attributes for each and every variant, which seems like nightmare. 

Maybe creating optional variant attributes would be good idea, for both - supporting existing structure of product variant attributes, and creating new more elastic attributes.

**Comment by maarcingebala** [comment]
Yes, we've discussed that recently and are considering introducing optional variant attributes to handle this. It will require some amount of changes in the core and in the storefront, as it's currently pretty hard assumption that every variant attribute's value needs to be defined.
